File: AES_dir/AES_test.cpp




#if 0
#include <openssl/aes.h>
#include "base64.h"


//https://blog.poxiao.me/p/advanced-encryption-standard-and-block-cipher-mode/

std::string fromhex(const char *str)
{
	size_t len = strlen(str) / 2;
	std::string strBuff;
	strBuff.resize(len);
	if (len > 512) len = 512;
	for (size_t i = 0; i < len; i++) {
		unsigned char c = 0;
		if (str[i * 2] >= '0' && str[i * 2] <= '9')
			c += (str[i * 2] - '0') << 4;
		if ((str[i * 2] & ~0x20) >= 'A' && (str[i * 2] & ~0x20) <= 'F')
			c += (10 + (str[i * 2] & ~0x20) - 'A') << 4;
		if (str[i * 2 + 1] >= '0' && str[i * 2 + 1] <= '9')
			c += (str[i * 2 + 1] - '0');
		if ((str[i * 2 + 1] & ~0x20) >= 'A' && (str[i * 2 + 1] & ~0x20) <= 'F')
			c += (10 + (str[i * 2 + 1] & ~0x20) - 'A');
		strBuff[i] = c;
	}
	return strBuff;
}

const char * hex_to_str( char *pSrc, int len)
{
	static char buf[512];
	for (int i = 0; i < len; i++){
		unsigned char cTmp = (unsigned char)pSrc[i];
		cTmp = (cTmp & 0xff);
		sprintf(buf + (i * 2), "%02x", cTmp);
	}
	return buf;
}

void test_main(int argc, char* argv[]) {
	std::string key = "hbbhehedadatttts";
	std::string plan_vec = "1234567890123456";
	std::vector<unsigned char> encode_vec(16);
	std::vector<unsigned char> decode_vec(16);

	// aes encrypt
	AES_KEY aes_enc_ctx;
	AES_set_encrypt_key( (const unsigned char *)key.data(), 128, &aes_enc_ctx);
	AES_ecb_encrypt((const unsigned char *)plan_vec.data(), encode_vec.data(), &aes_enc_ctx , AES_ENCRYPT);

	// aes decrypt
	AES_KEY aes_dec_ctx;
	AES_set_decrypt_key((const unsigned char *)key.data(), 128, &aes_dec_ctx);
	AES_ecb_encrypt(encode_vec.data(), decode_vec.data(), &aes_dec_ctx, AES_DECRYPT);

	std::string encoded_str = base64_encode((unsigned char const*)encode_vec.data(), encode_vec.size() );

	std::cout << "C++:"<< std::endl;
	std::cout << "源数据\t\t:" << plan_vec << std::endl;
	std::cout << "加密数据\t:" << encoded_str << std::endl;
	std::cout << "解密数据\t:" << std::string( (char *)decode_vec.data() , decode_vec.size() ) << std::endl;

}

#endif



#if 1

#include <boost/crc.hpp>


void test_main(int argc, char* argv[]) {
	std::string str_ = "womemjintianqushangxuezhou";
	boost::crc_32_type crc32;
	crc32.process_bytes(str_.data(), str_.size() );
	unsigned int ret = crc32.checksum();
	std::cout << "C++:" << std::endl;
	std::cout << "源数据\t\t:" << str_ << std::endl;
	std::cout << "校验值\t\t:" << ret << std::endl;
}

#endif

File: AES_dir/AES_test.py




def aes_encrypt(str_key , str_): 
	from Crypto.Cipher import AES
	obj = AES.new(str_key , AES.MODE_ECB)
	ciphertext = obj.encrypt(str_)
	result = base64.b64encode(ciphertext)  #base64 encode
	return result 
	
def aes_decode(str_key , str_base64):
	from Crypto.Cipher import AES
	str_data = base64.b64decode(str_base64)
	obj2 = AES.new(str_key , AES.MODE_ECB)
	return  obj2.decrypt(str_data)
	
def test_AES():
	str_key = 'hbbhehedadatttts'
	str_msg = '1234567890123456'
	str_ = aes_encrypt(str_key , str_msg)
	str_2 = aes_decode(str_key ,str_)
	print 'python'
	print "源数据\t\t:".decode("utf").encode("gbk"),str_msg
	print "加密数据\t:".decode("utf").encode("gbk"),str_
	print "解密数据\t:".decode("utf").encode("gbk"),str_2
	
	
def test_CRC():
	from zlib import crc32
	str_ = "womemjintianqushangxuezhou"
	ret_ = crc32(str_)
	print 'python'
	print "源数据\t\t:".decode("utf").encode("gbk"),str_
	print "校验值\t\t:".decode("utf").encode("gbk"),(ret_ & 0xFFFFFFFF )